src/util.cc

```diff
@@ -622,3 +624,46 @@ bool Truncate(const string& path, size_t size, string* err) {
   }
   return true;
 }
+
+bool ReplaceContent(const string& file_dst, const string& new_content,
+                    string* err) {
+#ifdef _WIN32
+  if (!DeleteFile(file_dst.c_str())) {
+    *err = GetLastErrorString();
+    return false;
+  }
+
+  if (rename(new_content.c_str(), file_dst.c_str()) < 0) {
+    *err = strerror(errno);
+    return false;
+  }
+#else
+  struct stat old_file;
+  bool found_uid_gid = true;
+
+  if (stat(file_dst.c_str(), &old_file) < 0) {
+    found_uid_gid = false;
+  }
+
+  if (unlink(file_dst.c_str()) < 0) {
+    *err = strerror(errno);
+    return false;
+  }
+
+  if (rename(new_content.c_str(), file_dst.c_str()) < 0) {
+    *err = strerror(errno);
+    return false;
+  }
+
+  // apply uid and gid again so we always stay as the uid and gid of the first
+  // ninja invocation that created the file at file_dst
+  if (found_uid_gid) {
+    if (chown(file_dst.c_str(), old_file.st_uid, old_file.st_gid)) {
+      Warning("Reapplying previous uid and gid failed with: %s uid: %d gid: %d",
```

Comment:
Do you know an example case where this might fail and how an error message might look like?

Reply:
I dont have an example, but i think we should tell the user one way or another. This is not returning false or causing any side effect changes to the pure nature of building.